### PR TITLE
Added ability to supply 'N' to Hilbert transform

### PR DESCRIPTION
**Pull request by choldgraf**
Mentioned in #2419. Rather than figuring out how to add zeros to a raw dataframe, I made some short tweaks to the Hilbert code. Think that's reasonable @Eric89GXL?

### CODE REVIEW

File: mne/io/base.py [hunk 1]
```diff
@@ -639,7 +639,8 @@ def apply_function(self, fun, picks, dtype, n_jobs, *args, **kwargs):
                 self._data[p, :] = data_picks_new[pp]
 
     @verbose
-    def apply_hilbert(self, picks, envelope=False, n_jobs=1, verbose=None):
+    def apply_hilbert(self, picks, envelope=False, n_jobs=1, verbose=None,
+                      N=None):
```
Comment (Eric89GXL) [inline mne/io/base.py]: `verbose` is always last

File: mne/io/fiff/tests/test_raw.py [hunk 2]
```diff
@@ -888,9 +888,16 @@ def test_hilbert():
...
 
     raw2 = raw.copy()
+    raw3 = raw.copy()
+    raw4 = raw.copy()
     raw.apply_hilbert(picks)
     raw2.apply_hilbert(picks, envelope=True, n_jobs=2)
 
+    # Test custom N
+    raw3.apply_hilbert(picks, N=raw3.n_times + 100)
+    assert_equal(raw3._data.shape, raw._data.shape)
```
Comment (Eric89GXL) [inline mne/io/fiff/tests/test_raw.py]: How closely do the resulting values match between these two? I'm not too familiar with edge effects due to hilbert transforms. Do they match up pretty well but degrade near the end / zero-padding area? Or are the values disturbed more uniformly everywhere?

Comment (Eric89GXL) [inline mne/io/fiff/tests/test_raw.py]: (it would be nice to be able to add a note saying what effect the zero-padding will have)

**Comment by Eric89GXL** [comment]
Looks like a reasonable approach to me. It would be nice if we could get some `assert_allclose` test on some subset of the zero-padded and non-zero-padded data if possible. It might not be, depending on how the zero-padding affects the signal, though.

File: mne/io/base.py (hunk 1 above)
Comment (agramfort) [inline mne/io/base.py]: N is a bad name. We use n_fft I think elsewhere, nfft ...

File: mne/io/fiff/tests/test_raw.py (hunk 2 above)
Comment (choldgraf) [inline mne/io/fiff/tests/test_raw.py]: Hmmm, ok so it does seem like the padding will affect the signal in more than just the peri-zero region. Does anyone in the MNE world have a more fundamental understanding of how padding and FFTs interact with each other?

e.g.:
```python
out1 = sg.hilbert(raw._data[0, :1000])
out2 = sg.hilbert(raw._data[0, :1000], N=1500)

print(raw3._data[:5, 0])
print(raw._data[:5, 0])
```
Output:
```python
[ -2.58713276e-12 +1.87655655e-11j   1.31091416e-12 -9.42899838e-12j
  -1.45324088e-14 -1.43578007e-13j  -2.77344749e-12 +2.45830856e-11j
   5.02734681e-13 +0.00000000e+00j]
[ -2.58713270e-12   1.31091411e-12  -1.45324088e-14  -2.77344739e-12
   5.02734681e-13]
```

Comment (choldgraf) [inline mne/io/fiff/tests/test_raw.py]: It seems like maybe this is not as big of an issue for bandpass filtered signals? Maybe the differences between them are within the tolerance that matters? Not sure how to check for this... e.g.:

```python
f, (ax1, ax2) = plt.subplots(2, 1, figsize=(15, 10))
raw.filter(20, 40)
out1 = sg.hilbert(raw._data[0, :1000])
out2 = sg.hilbert(raw._data[0, :1000], N=1500)[:1000]
out3 = sg.hilbert(raw._data[0, :1000], N=5000)[:1000]

for dat in [out1, out2, out3]:
    print(dat[:5])
    print(np.abs(dat[:5]))
    print('\n')
    ax1.plot(np.abs(dat[:200]))
    ax1.plot(dat[:200])
    ax2.plot(np.abs(dat[-200:]))
    ax2.plot(dat[-200:])
    
```


image

Comment (agramfort) [inline mne/io/fiff/tests/test_raw.py]: if you high pass the data or use a window it shouldn't make a big
difference. When you pad you are basically doing spectrum oversampling yet
using circular convolutions (hence the edge artifacts spreading)

Comment (choldgraf) [inline mne/io/fiff/tests/test_raw.py]: OK so how do you think this should be incorporated into the tests then? Just allow for a larger difference in the call to `allclose`? And maybe include a line in the docstring that zero padding will change things but not too much for bandpassed signals?

Comment (Eric89GXL) [inline mne/io/fiff/tests/test_raw.py]: sounds reasonable to me

Comment (agramfort) [inline mne/io/fiff/tests/test_raw.py]: Ok for me

**Comment by choldgraf** [comment]
OK, how about this?

File: mne/io/base.py [hunk 3]
```diff
@@ -688,12 +693,21 @@ def apply_hilbert(self, picks, envelope=False, n_jobs=1, verbose=None):
         MNE inverse solution, the enevlope in source space can be obtained
         by computing the analytic signal in sensor space, applying the MNE
         inverse, and computing the envelope in source space.
+
+        Also note that the n_fft parameter will allow you to pad the signal
```
Comment (Eric89GXL) [inline mne/io/base.py]: pad the end of the signal

Comment (Eric89GXL) [inline mne/io/base.py]: (unless you symmetrically pad)

Comment (Eric89GXL) [inline mne/io/base.py]: but actually it shouldn't matter if your padding is symmetrical or not, since the FFT is circular, so never mind!

**Comment by Eric89GXL** [comment]
LGTM, @agramfort feel free to merge if you're happy too

**Comment by agramfort** [comment]
Thanks

**Comment by choldgraf** [comment]
cool, thanks guys

**Pull request merged**